### PR TITLE
Simple websocket server

### LINKED ISSUE #14: Create basic websocket server
It's time to write code!
We need to understand the [Boost.Beast examples](https://www.boost.org/doc/libs/1_72_0/libs/beast/doc/html/beast/examples.html#beast.examples.servers) and write a simple websocket server. I suggest writing an asynchronous version right away. The listening address and the port will be passed through command line arguments.

The server main goals:
- Accept connections and create sessions;
- Handle sessions for reading and writing data to client.

We also need to choose what we will base on: asynchronous code or coroutine code.

@c71n93 @temikfart any additions?

### PR DESCRIPTION
**Pull request by cann3v**
Stackful coroutine based websocket server.
Executable takes 3 arguments: address, port, threads. For example `./coolstory-gram-server 0.0.0.0 8080 2`.
This arguments passed into `void run(...)`, within which an event loop is started (`ioc.run()`) for the specified number of threads.
Function `do_listen(...)` accepts tcp connections and spawns `do_accept(...)` for each connection.
Function `do_accept(...)` performs some actions:
- Accepts the handshake;
- Reads from socket and echoes back while connection is open.

### CODE REVIEW

File: include/CSGServer/websocket_session.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+namespace beast = boost::beast;         // from <boost/beast.hpp>
+namespace http = beast::http;           // from <boost/beast/http.hpp>
+namespace websocket = beast::websocket; // from <boost/beast/websocket.hpp>
+namespace net = boost::asio;            // from <boost/asio.hpp>
+using tcp = boost::asio::ip::tcp;       // from <boost/asio/ip/tcp.hpp>
+
+void fail(beast::error_code ec, char const* what);
+void do_session(websocket::stream<beast::tcp_stream>& ws, net::yield_context yield);
+void do_listen(net::io_context& ioc, tcp::endpoint endpoint, net::yield_context yield);
+void run(net::ip::address address, unsigned short port, int threads);
```
Comment (temikfart) [inline include/CSGServer/websocket_session.hpp]: Add new empty line at the end of the file.

File: src/main.cpp [hunk 2]
```diff
@@ -1,7 +1,22 @@
 #include "CSGServer/main.hpp"
 
-int main() {
-    std::cout << "Hello, world!" << std::endl;
+int main(int argc, char** argv) {
+    logger::init(logger::debug, logger::cout);
 
-    return 0;
+    // Check command line arguments.
```
Comment (temikfart) [inline src/main.cpp]: This is an extra comment. Do not write comments like this one, if it's a simple thing in code.

File: src/main.cpp [hunk 3]
```diff
@@ -1,7 +1,22 @@
...
 
-int main() {
-    std::cout << "Hello, world!" << std::endl;
+int main(int argc, char** argv) {
+    logger::init(logger::debug, logger::cout);
 
-    return 0;
+    // Check command line arguments.
+    if (argc != 4)
+    {
```
Comment (temikfart) [inline src/main.cpp]: According to [Codestyle](https://github.com/CoolStory-Studio/coolstory-gram/discussions/2), the curly brace must be in the same line as the operator `if`.

File: src/main.cpp [hunk 4]
```diff
@@ -1,7 +1,22 @@
...
+        std::cerr <<
+                  "Usage: coolstory_gram_server <address> <port> <threads>\n" <<
+                  "Example:\n" <<
+                  "    coolstory_gram_server 0.0.0.0 8080 1\n";
+        return EXIT_SUCCESS;
```
Comment (temikfart) [inline src/main.cpp:11]: `EXIT_FAILURE` more preferable.

File: src/main.cpp [hunk 5]
```diff
@@ -1,7 +1,22 @@
...
+    }
+    auto const address = net::ip::make_address(argv[1]);
+    auto const port = static_cast<unsigned short>(std::atoi(argv[2]));
```
Comment (temikfart) [inline src/main.cpp:21]: Not for changes: _If you don't have any specific cause to use unusual type, than use `int`._

File: src/websocket_session.cpp [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            [](websocket::response_type& res)
+            {
+                res.set(http::field::server,
+                        std::string(BOOST_BEAST_VERSION_STRING) +
+                        " coolstory-gram-server");
+            }));
+
+    // Accept the websocket handshake
+    ws.async_accept(yield[ec]);
+    if(ec)
```
Comment (temikfart) [inline src/websocket_session.cpp]: Spaces!

File: src/websocket_session.cpp [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        return fail(ec, "accept");
+
+    for(;;)
+    {
```
Comment (temikfart) [inline src/websocket_session.cpp]: Spaces!

File: src/websocket_session.cpp [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        // This buffer will hold the incoming message
+        beast::flat_buffer buffer;
+
+        // Read a message
+        ws.async_read(buffer, yield[ec]);
+
+        // This indicates that the session was closed
+        if(ec == websocket::error::closed) {
```
Comment (temikfart) [inline src/websocket_session.cpp]: Spaces!

File: src/websocket_session.cpp [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+            LOGD << "Client closed the connection";
+            break;
+        }
+
+        if(ec)
```
Comment (temikfart) [inline src/websocket_session.cpp]: Spaces!

File: src/websocket_session.cpp [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+            return fail(ec, "read");
+
+        LOGD << "(" << ws.next_layer().socket().remote_endpoint() << "): " << beast::make_printable(buffer.data());
+
+        // Echo the message back
+        ws.text(ws.got_text());
+        ws.async_write(buffer.data(), yield[ec]);
+        if(ec)
```
Comment (temikfart) [inline src/websocket_session.cpp]: Spaces!

File: src/websocket_session.cpp [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    }
+}
+
+void do_listen(net::io_context& ioc, tcp::endpoint endpoint, net::yield_context yield) {
+    beast::error_code ec;
+
+    // Open the acceptor
+    tcp::acceptor acceptor(ioc);
+    acceptor.open(endpoint.protocol(), ec);
+    if(ec)
```
Comment (temikfart) [inline src/websocket_session.cpp]: Spaces!

File: src/websocket_session.cpp [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    // Start listening for connections
+    acceptor.listen(net::socket_base::max_listen_connections, ec);
+    if(ec)
+        return fail(ec, "listen");
+
+    for(;;)
+    {
+        tcp::socket socket(ioc);
+        acceptor.async_accept(socket, yield[ec]);
+        if(ec)
```
Comment (temikfart) [inline src/websocket_session.cpp]: Spaces!

File: src/websocket_session.cpp [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+            fail(ec, "accept");
+        else
+            LOGD << "Connection accepted from " << socket.remote_endpoint();
+            boost::asio::spawn(
+                    acceptor.get_executor(),
+                    std::bind(
+                            &do_session,
+                            websocket::stream<
+                                    beast::tcp_stream>(std::move(socket)),
+                            std::placeholders::_1));
```
Comment (temikfart) [inline src/websocket_session.cpp]: `else` branch must be with curly braces or fix tabulation.

File: src/websocket_session.cpp [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+    // Run the I/O service on the requested number of threads
+    std::vector<std::thread> v;
+    v.reserve(threads - 1);
+    for(auto i = threads - 1; i > 0; --i)
+        v.emplace_back(
+                [&ioc]
+                {
+                    ioc.run();
+                });
```
Comment (temikfart) [inline src/websocket_session.cpp]: put code in `for` body using curly braces.

Comment (temikfart) [reply src/websocket_session.cpp]: And fix spacing.

**Review by temikfart — CHANGES_REQUESTED** [review]
Mostly codestyle issues, but not all.

File: src/websocket_session.cpp [hunk 15]
```diff
@@ -0,0 +1,120 @@
+#include "CSGServer/websocket_session.hpp"
+
+void fail(beast::error_code ec, char const* what) {
+    LOGF << what << ": " << ec.message();
+}
```
Comment (temikfart) [inline src/websocket_session.cpp:5]: Does this function should do `exit()`?

Comment (cann3v) [reply src/websocket_session.cpp:5]: No, it doesn't. When the application is serving the client, an error may occur (connection, read/write websocket etc). In this case, the connection with client is closed, but the application continues to serve other clients.

**Review by temikfart — APPROVED** [review]

**Pull request merged**